`npm test` fails with authentication error

I installed and am running strato2 on localhost:3000. When running `npm test`, I get a 401 response (included below). @amcgee, I'm still investigating this, but on the surface it makes sense to me that this would fail, because I don't see where any authentication token is given when the request is made (`test/gateway/_scaffold_.gateway.test.js:70`). I may be missing something, though.

      Gateway 'http'
    Creating monitor...
        1) "before all" hook
    
      0 passing (117ms)
      1 failing
    
      1) Gateway 'http' "before all" hook:
    
          Uncaught Error: expected 401 to sort of equal 201
          + expected - actual
    
          -401
          +201
          
          at test/gateway/_scaffold_.gateway.test.js:79:33
          at _stream_readable.js:908:16
    
    
    
    npm ERR! Test failed.  See above for more details.

--- a/.travis.yml
+++ b/.travis.yml
@@ -1,0 +1,8 @@
+language: node_js
+node_js:
+  - '0.12'
+services:
+  - mongodb
+notifications:
+  slack: welldonetech:yF3HybtHHNfxzf0nKhEyG0mC
+sudo: false

--- a/README.md
+++ b/README.md
@@ -1,18 +1,17 @@
+[![Build Status](https://travis-ci.org/WellDone/strato2.svg)](https://travis-ci.org/WellDone/strato2)
+
+[![Dependencies](https://david-dm.org/WellDone/strato2.svg)](https://david-dm.org/WellDone/strato2)
+
 Strato Data portal
 =======
 A refresh of the original Strato ([welldone/strato](http://github.com/WellDone/Strato)), this time 100x cleaner and 10x better.
 
-# Installation and Usage (Ubuntu)
+# Development Environment
 
-```shell
-curl -sL https://deb.nodesource.com/setup | sudo bash -
-sudo apt-get install -y git nodejs
-sudo npm install -g bower gulp
-
-git clone https://www.github.com/welldone/strato2
-cd strato2
-npm install
-
-nohup npm start &
-npm test
+```console
+$ vagrant up
+$ vagrant ssh
+$ npm install
+$ npm test
+$ npm start
 ```

--- a/Vagrantfile
+++ b/Vagrantfile
@@ -9,7 +9,8 @@ Vagrant.configure(VAGRANTFILE_API_VERSION) do |config|
 
   # config.vm.box_check_update = false
 
-  config.vm.network "forwarded_port", guest: 3000, host: 8080
+  config.vm.network "forwarded_port", guest: 3000, host: (ENV["WEB_PORT"] || 3000)
+  config.vm.network "forwarded_port", guest: 27017, host: (ENV["MONGO_PORT"] || 27017)
 
   # config.vm.synced_folder "../data", "/vagrant_data"
 

--- a/gulpfile.js
+++ b/gulpfile.js
@@ -1,25 +1,16 @@
 var gulp = require('gulp');
 var handlebars = require('gulp-compile-handlebars');
 var rename = require('gulp-rename');
-var clean = require('gulp-clean');
+var del = require('del');
 var bower = require('bower');
 
-var watch = false;
-process.argv.forEach(function (val, index, array) {
-  if ( val == '-w' || val == '--watch' )
-  {
-    watch = true;
-  }
-});
-
 gulp.task('clean', function(){
-  return gulp.src(['www/*'], {read:false})
-    .pipe(clean());
+  return del(['www/*']);
 });
 
 gulp.task('bower', function(cb){
   bower.commands.install([], {save: true}, {})
-    .on('end', function(installed){
+    .on('end', function(){
       cb(); // notify gulp that this task is finished
     });
 });
@@ -36,17 +27,17 @@ gulp.task('hbs', function () {
     options = {
         batch : ['./src/html/_partials'],
         helpers : {}
-    }
+    };
 
     return gulp.src(['src/html/**/*.hbs', '!src/html/_partials/**/*.*'])
         .pipe(handlebars(templateData, options))
         .pipe(rename(function (path) {
-            path.extname = ".html"
+            path.extname = '.html';
         }))
         .pipe(gulp.dest('./www'));
-})
+});
 
 gulp.task('default', ['bower','move', 'hbs']);
 gulp.task('watch', ['default'],function() {
-  var watcher = gulp.watch('src/**/*.*', ['default']);
-})
+  gulp.watch('src/**/*.*', ['default']);
+});

--- a/index.js
+++ b/index.js
@@ -1,21 +1,9 @@
-var BPromise = require('bluebird');
+var strato = require('./lib/strato');
 
+if ( !process.env.DATABASE_URL )
+{
+	console.error("Environment variable DATABASE_URL does not exist.");
+	process.exit(-1);
+}
 
-var api = require('./lib/api/api.v1.0');
-var db = require('./lib/db');
-
-console.log("Connecting to the database...");
-BPromise.join(
-	api.connect(),
-	db.connect()
-).then(function() {
-	var server = require('./lib/server');
-	
-	var port = process.env['PORT'] || 3000;
-	console.log("Starting server on port %d...", port);
-	server.listen( port );
-
-	// if ( !db.db.collection('people').findOne({masterOfTheUniverse: true}) ) {
-	// 	console.error("WARNING: No master of the universe!")
-	// }
-})
+strato.start({ db: process.env.DATABASE_URL, port: process.env.PORT || 3000 });

--- a/lib/api-router.js
+++ b/lib/api-router.js
@@ -1,8 +1,10 @@
 var Router = require('./util/router');
 
-var routes = {
-	'v1': require( './api/api.v1.0' ).router,
-	'v1.0': require( './api/api.v1.0' ).router
-}
+module.exports = function(api) {
+	var routes = {
+		'v1': api.router,
+		'v1.0': api.router
+	};
 
-module.exports = Router(routes);
+	return Router(routes);
+};

--- a/lib/api/api.v1.0.js
+++ b/lib/api/api.v1.0.js
@@ -1,84 +1,85 @@
 var REM = require('remjs');
-var Router = require('../util/router');
 
-module.exports = new REM( {
-	version: "1.0",
-	engine: REM.engine.mongodb({
-		url: process.env.DATABASE_URL
-	}),
-	authentication: {
-		annonymous_signup: true,
-		login_authority: { 
-			resource: 'people'
-		}
-	},
-	permissions: {
-		defaults: {
-			'create': function( id ) { return id.masterOfTheUniverse ? true : false; },
-			'read': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
-			'update': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
-			'delete': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; }
-		},
-		annonymous: []
-	},
-	resources: {
-		'monitors': {
-			children: ['reports']
-		},
-		'reports': {
-			forbid: ['post']
-		},
-		'webhooks': {
-			permissions: {
-				defaults: {
-					'create': function( id ) {
-						if ( id.masterOfTheUniverse || id.organization )
-							return true;
-						else
-							return false;
-					},
-					'read': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
-					'update': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
-					'delete': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; }
-				}
-			},
-			defaults: {
-				owner: function(resource, req) {
-					return req.rem.identity.organization;
-				}
+module.exports = function(opts) {
+	return new REM( {
+		version: "1.0",
+		engine: REM.engine.mongodb({
+			url: opts.db
+		}),
+		authentication: {
+			annonymous_signup: true,
+			login_authority: { 
+				resource: 'people'
 			}
 		},
-		'people': {
-			immutable_keys: ['username', 'masterOfTheUniverse', 'organization'],
-			permissions: {
+		permissions: {
+			defaults: {
+				'create': function( id ) { return id.masterOfTheUniverse ? true : false; },
+				'read': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
+				'update': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
+				'delete': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; }
+			},
+			annonymous: []
+		},
+		resources: {
+			'monitors': {
+				children: ['reports']
+			},
+			'reports': {
+				forbid: ['post']
+			},
+			'webhooks': {
+				permissions: {
+					defaults: {
+						'create': function( id ) {
+							if ( id.masterOfTheUniverse || id.organization )
+								return true;
+							else
+								return false;
+						},
+						'read': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
+						'update': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; },
+						'delete': function( id ) { return id.masterOfTheUniverse ? true : { $and: [{owner: {$exists: true}}, { owner: id.organization }] }; }
+					}
+				},
 				defaults: {
-					'create': function( id ) {
-						return id.masterOfTheUniverse ? true : false;
-					},
-					'read': function( id ) {
-						if ( id.masterOfTheUniverse )
-							return true;
-						else
-							return { $or: [ 
-								{ _id: id._id }, 
-								{ $and: [
-									{ organization: {$exists: true}},
-									{ organization: id.organization }
-								] }
-							] };
-					},
-					'update': function( id ) {
-						if ( id.masterOfTheUniverse )
-							return true;
-						else
-							return { _id: id._id };
-					},
-					'delete': function( id ) {
-						if ( id.masterOfTheUniverse )
-							return { _id: {$ne: id._id } };
+					owner: function(resource, req) {
+						return req.rem.identity.organization;
+					}
+				}
+			},
+			'people': {
+				immutable_keys: ['username', 'masterOfTheUniverse', 'organization'],
+				permissions: {
+					defaults: {
+						'create': function( id ) {
+							return id.masterOfTheUniverse ? true : false;
+						},
+						'read': function( id ) {
+							if ( id.masterOfTheUniverse )
+								return true;
+							else
+								return { $or: [ 
+									{ _id: id._id }, 
+									{ $and: [
+										{ organization: {$exists: true}},
+										{ organization: id.organization }
+									] }
+								] };
+						},
+						'update': function( id ) {
+							if ( id.masterOfTheUniverse )
+								return true;
+							else
+								return { _id: id._id };
+						},
+						'delete': function( id ) {
+							if ( id.masterOfTheUniverse )
+								return { _id: {$ne: id._id } };
+						}
 					}
 				}
 			}
 		}
-	}
-} );
+	} );
+};

--- a/lib/db.js
+++ b/lib/db.js
@@ -11,15 +11,10 @@ var DB = function() {
 
 DB.prototype.collection = function(name) {
 	return this.db.collection(name);
-}
+};
 
-DB.prototype.connect = function() {
-	if ( !process.env.DATABASE_URL )
-	{
-		console.error("Environment variable DATABASE_URL does not exist.");
-		process.exit(-1);
-	}
-	return MongoClient.connectAsync( process.env.DATABASE_URL )
+DB.prototype.connect = function(dbUrl) {
+	return MongoClient.connectAsync( dbUrl )
 		.bind(this)
 		.then(function(db) {
 			this.db = db;

--- a/lib/server.js
+++ b/lib/server.js
@@ -1,28 +1,30 @@
 var express = require( 'express' );
 
-var app = express();
+module.exports = function(api) {
+	var app = express();
 
-var bodyParser = require( 'body-parser' )
-var compression = require( 'compression' );
+	var bodyParser = require( 'body-parser' )
+	var compression = require( 'compression' );
 
-var path = require('path');
+	var path = require('path');
 
-// parse application/x-www-form-urlencoded
-app.use(bodyParser.urlencoded({ extended: false }))
+	// parse application/x-www-form-urlencoded
+	app.use(bodyParser.urlencoded({ extended: false }))
 
-// parse application/json
-app.use(bodyParser.json())
+	// parse application/json
+	app.use(bodyParser.json())
 
-// parse text
-app.use(bodyParser.text())
+	// parse text
+	app.use(bodyParser.text())
 
-app.use( compression({
-  threshold: 512
-}) );
+	app.use( compression({
+	  threshold: 512
+	}) );
 
-app.use('/api', require('./api-router') );
-app.use('/gateway', require('./gateway-router') );
+	app.use('/api', require('./api-router')(api) );
+	app.use('/gateway', require('./gateway-router') );
 
-app.use("/", express.static(path.join( __dirname, '..', 'www' )));
+	app.use("/", express.static(path.join( __dirname, '..', 'www' )));
 
-module.exports = app;
+	return app;
+}

--- a/lib/strato.js
+++ b/lib/strato.js
@@ -1,20 +1,18 @@
 var BPromise = require('bluebird');
-var api = require('./api/api.v1.0');
 var db = require('./db');
 
 module.exports.start = function(params) {
-	console.log("Connecting to the database...");
-	BPromise.join(
+	console.log("Connecting to the database '%s'...", params.db);
+
+	var api = require('./api/api.v1.0')(params);
+
+	return BPromise.join(
 		api.connect(),
 		db.connect(params.db)
 	).then(function() {
-		var server = require('./server');
+		var server = BPromise.promisifyAll(require('./server')(api));
 
 		console.log("Starting server on port %d...", params.port);
-		server.listen(params.port);
-
-		// if ( !db.db.collection('people').findOne({masterOfTheUniverse: true}) ) {
-		// 	console.error("WARNING: No master of the universe!")
-		// }
+		return server.listen(params.port);
 	});
 };

--- a/lib/strato.js
+++ b/lib/strato.js
@@ -1,0 +1,20 @@
+var BPromise = require('bluebird');
+var api = require('./api/api.v1.0');
+var db = require('./db');
+
+module.exports.start = function(params) {
+	console.log("Connecting to the database...");
+	BPromise.join(
+		api.connect(),
+		db.connect(params.db)
+	).then(function() {
+		var server = require('./server');
+
+		console.log("Starting server on port %d...", params.port);
+		server.listen(params.port);
+
+		// if ( !db.db.collection('people').findOne({masterOfTheUniverse: true}) ) {
+		// 	console.error("WARNING: No master of the universe!")
+		// }
+	});
+};

--- a/lib/util/report-processor.js
+++ b/lib/util/report-processor.js
@@ -36,7 +36,7 @@ var addReport = function( report, options, callback ) {
 						})
 					})
 				});
-		callback();
+		callback(null, report);
 	} );
 	db.collection('monitors').update({_id: options.monitors_id}, { $set: { status: options.status } }, { multi: false }, function (err, numReplaced) {
 		console.log(err, numReplaced);

--- a/lib/util/users.js
+++ b/lib/util/users.js
@@ -1,0 +1,30 @@
+var rp = require('request-promise');
+
+var Promise = require('bluebird');
+var mongodb = Promise.promisifyAll(require('mongodb'));
+Promise.promisifyAll(mongodb.Cursor.prototype);
+
+var MongoClient = Promise.promisifyAll(mongodb.MongoClient);
+
+module.exports = function(opts) {
+	return {
+		create: function(username, password) {
+			return rp.post('http://localhost:' + opts.port + '/api/v1/_signup')
+			.form({ login: username, password: password });
+		},
+
+		update: function(username, properties) {
+			return MongoClient.connectAsync( opts.db )
+			.then(function(db) {
+				var collection = db.collection('people');
+				return collection.updateAsync(
+					{ username: username },
+					{ $set: properties }, 
+					{})
+				.then(function() {
+					return db.close();
+				});
+			});
+		}
+	};
+};

--- a/package.json
+++ b/package.json
@@ -36,5 +36,9 @@
   "devDependencies": {
     "mocha": "^2.0.1",
     "expect.js": "^0.3.1"
+  },
+  "engines": {
+    "node": "0.12.x",
+    "npm": "2.x"
   }
 }

--- a/package.json
+++ b/package.json
@@ -26,7 +26,7 @@
     "gulp-compile-handlebars": "^0.3.3",
     "gulp": "^3.8.10",
     "gulp-rename": "^1.2.0",
-    "gulp-clean": "^0.3.1",
+    "del": "^2.0.0",
     "bower": "^1.3.12",
     "handlebars": "^2.0.0",
     "supervisor": "^0.6.0",

--- a/package.json
+++ b/package.json
@@ -35,7 +35,9 @@
   },
   "devDependencies": {
     "mocha": "^2.0.1",
-    "expect.js": "^0.3.1"
+    "expect.js": "^0.3.1",
+    "assert": "^1.3.0",
+    "request-promise": "^0.4.3"
   },
   "engines": {
     "node": "0.12.x",

--- a/scripts/provision.sh
+++ b/scripts/provision.sh
@@ -1,39 +1,15 @@
-apt-get update
-apt-get install -y nginx git
+set -e
+set -x
 
-apt-get install -y python-software-properties python g++ make curl
-curl -sL https://deb.nodesource.com/setup | sudo bash -
-apt-get update
-apt-get install -y nodejs
+apt-get -qq update
+apt-get -qq install -y python-software-properties python g++ make curl git
+curl -sL https://deb.nodesource.com/setup_0.12 | sudo bash -
 
 sudo apt-key adv --keyserver hkp://keyserver.ubuntu.com:80 --recv 7F0CEB10
 echo 'deb http://downloads-distro.mongodb.org/repo/ubuntu-upstart dist 10gen' | sudo tee /etc/apt/sources.list.d/mongodb.list
-sudo apt-get update
-sudo apt-get install -y mongodb-org
-
+sudo apt-get -qq update
+sudo apt-get -qq install -y nodejs mongodb-org
 
 echo "export DATABASE_URL=mongodb://localhost:27017/strato" >> /home/vagrant/.profile
 
-if [ "$1" == "production" ]; then
-	git clone https://www.github.com/welldone/strato2.git /welldone
-else
-	rm /welldone || true
-	if [ -d "/vagrant" ]; then
-		ln -s /vagrant /welldone
-	fi
-fi
-
-if [ ! -d "/welldone" ]; then
-	echo "No /welldone directory found."
-	exit 1
-fi
-
-cd /welldone
-npm install
-
-ln -fs /welldone/scripts/upstart.conf /etc/init/strato.conf
-initctl reload-configuration
-start strato
-
-ln -fs /welldone/scripts/nginx.conf /etc/nginx/sites-enabled/default
-sudo /etc/init.d/nginx restart
+echo "cd /vagrant" >> /home/vagrant/.bashrc

--- a/test/_scaffold_.test.js
+++ b/test/_scaffold_.test.js
@@ -1,0 +1,65 @@
+var strato = require('../lib/strato');
+var assert = require('assert');
+var artifacts = require('./test-artifacts');
+var superagent = require('superagent');
+
+var Promise = require('bluebird');
+var mongodb = Promise.promisifyAll(require('mongodb'));
+Promise.promisifyAll(mongodb.Cursor.prototype);
+
+var MongoClient = Promise.promisifyAll(mongodb.MongoClient);
+
+before(function() {
+	this.timeout(5000);
+
+	console.log('Constructing test scaffolding...');
+	var users = require('../lib/util/users.js')({ db: artifacts.db, port: artifacts.port });
+
+	return strato.start(artifacts)
+	.then(function() {
+		console.log('Creating test user...');
+		return users.create(artifacts.username, artifacts.password);
+	})
+	.then(function() {
+		console.log('Granting test user master permissions...');
+		return users.update(artifacts.username, { masterOfTheUniverse: true });
+	})
+	.then(function() {
+		console.log('Log in and obtain user token...');
+
+		return new Promise(function(resolve, reject) {
+			superagent
+			.post(artifacts.url + '/api/v1/_login')
+			.send( {
+				login: artifacts.username,
+				password: artifacts.password
+			})
+			.end(function(e, res) {
+				if (e) {
+					reject(e);
+				}
+				else if (res.status !== 200) {
+					reject(new Error(res.status + ': ' + res.text));
+				}
+				else {
+					artifacts.token = res.text;
+					resolve();
+				}
+			});
+		});
+	})
+	.then(function() {
+		console.log('...scaffolding constructed.');
+	});
+});
+
+after(function() {
+	this.timeout(5000);
+
+	console.log('Cleaning up database %s', artifacts.db);
+
+	return MongoClient.connectAsync(artifacts.db)
+	.then(function(db) {
+		return db.dropDatabase();
+	});
+});

--- a/test/gateway/_scaffold_.gateway.test.js
+++ b/test/gateway/_scaffold_.gateway.test.js
@@ -1,6 +1,7 @@
 var superagent = require('superagent')
 var expect = require('expect.js')
 var _ = require('lodash')
+var artifacts = require('../test-artifacts');
 
 function ConstructReport( version ) {
   var bytes;
@@ -67,7 +68,8 @@ function testGateway( gatewayName, minVersion, payloadFormatter, callback ) {
     var gsmid = "+5555555";
     before(function(done){
       console.log("Creating monitor...");
-      superagent.post('http://localhost:3000/api/v1/monitors')
+      superagent.post(artifacts.url + '/api/v1/monitors')
+        .set('Authorization', 'Bearer ' + artifacts.token)
         .send({
           name: 'TestMonitor',
           location: [0,0],
@@ -97,7 +99,7 @@ function testGateway( gatewayName, minVersion, payloadFormatter, callback ) {
             content: report
           }
           payload = payloadFormatter(payload);
-          superagent.post('http://localhost:3000/gateway/' + gatewayName)
+          superagent.post(artifacts.url + '/gateway/' + gatewayName)
             .type(typeof payload == 'string'? 'text':'json')
             .send(payload)
             .end(function(e,res){
@@ -108,7 +110,7 @@ function testGateway( gatewayName, minVersion, payloadFormatter, callback ) {
         })
 
         it('get the report to ensure it was created', function(done) {
-          superagent.get('http://localhost:3000/api/v1/monitors/' + monitor._id + '/reports')
+          superagent.get(artifacts.url + '/api/v1/monitors/' + monitor._id + '/reports')
             .end(function(e,res){
               expect(e).to.eql(null);
               expect(res.status).to.eql(200);
@@ -121,7 +123,7 @@ function testGateway( gatewayName, minVersion, payloadFormatter, callback ) {
             })
         })
         it('clean up the report', function(done) {
-          superagent.del('http://localhost:3000/api/v1/monitors/' + monitor._id + '/reports/' + reportID)
+          superagent.del(artifacts.url + '/api/v1/monitors/' + monitor._id + '/reports/' + reportID)
             .end(function(e,res){
               expect(e).to.eql(null);
               expect(res.status).to.eql(200);
@@ -133,7 +135,8 @@ function testGateway( gatewayName, minVersion, payloadFormatter, callback ) {
     
     after(function(done){
       console.log("Deleting monitor...");
-      superagent.del('http://localhost:3000/api/v1/monitors/' + monitor._id)
+      superagent.del(artifacts.url + '/api/v1/monitors/' + monitor._id)
+        .set('Authorization', 'Bearer ' + artifacts.token)
         .end(function(e,res){
           expect(e).to.eql(null);
           expect(res.status).to.eql(200);

--- a/test/gateway/smssync.gateway.test.js
+++ b/test/gateway/smssync.gateway.test.js
@@ -5,7 +5,7 @@ gatewayScaffold( 'smssync', 2, function(payload) {
   return {
     message: payload.content,
     from: payload.gsmid,
-    send_timestamp: payload.timestamp.getTime()
+    sent_timestamp: payload.timestamp.getTime()
   }
 }, function(payload, res) {
   expect(res.status).to.eql(200);

--- a/test/monitors.api.test.js
+++ b/test/monitors.api.test.js
@@ -3,17 +3,21 @@ var expect = require('expect.js')
 var _ = require('lodash')
 
 describe('REST api v1.0, monitors resource:', function(){
+  var artifacts = require('./test-artifacts');
+
   var id = null;
   var name = 'TestMonitor'
   var location = [0,0]
 
   it('create new monitor', function(done){
-    superagent.post('http://localhost:3000/api/v1/monitors')
+    superagent.post(artifacts.url + '/api/v1/monitors')
+      .set('Authorization', 'Bearer ' + artifacts.token)
       .send({
       	name: name,
         location: location
       })
       .end(function(e,res){
+        console.log(res.text);
       	expect(e).to.eql(null);
       	expect(res.status).to.eql(201);
         expect(res.body).to.be.an('object');
@@ -24,7 +28,8 @@ describe('REST api v1.0, monitors resource:', function(){
       })
   });
   it('fetch new monitor', function(done){
-  	superagent.get('http://localhost:3000/api/v1/monitors/' + id)
+  	superagent.get(artifacts.url + '/api/v1/monitors/' + id)
+      .set('Authorization', 'Bearer ' + artifacts.token)
   		.end(function(e,res){
   			expect(e).to.eql(null);
   			expect(res.status).to.eql(200);
@@ -38,7 +43,7 @@ describe('REST api v1.0, monitors resource:', function(){
   })
 
   it('get the empty reports list', function(done) {
-  	superagent.get('http://localhost:3000/api/v1/monitors/' + id + '/reports')
+  	superagent.get(artifacts.url + '/api/v1/monitors/' + id + '/reports')
   		.end(function(e,res){
   			expect(e).to.eql(null);
   			expect(res.status).to.eql(200);
@@ -48,10 +53,11 @@ describe('REST api v1.0, monitors resource:', function(){
   			done()
   		})
   })
+
   var reports = [];
   _.forEach( ['a','b','c'], function(dummy_value) {
   	it('post a report', function(done){
-	  	superagent.post('http://localhost:3000/api/v1/monitors/' + id + '/reports')
+	  	superagent.post(artifacts.url + '/api/v1/monitors/' + id + '/reports')
 	  		.send({
 	  			value: dummy_value
 	  		})
@@ -70,7 +76,7 @@ describe('REST api v1.0, monitors resource:', function(){
   })
 
   it('check that all the reports show up', function(done) {
-  	superagent.get('http://localhost:3000/api/v1/monitors/' + id + '/reports')
+  	superagent.get(artifacts.url + '/api/v1/monitors/' + id + '/reports')
   		.end(function(e,res){
   			expect(e).to.eql(null);
   			expect(res.status).to.eql(200);
@@ -87,7 +93,7 @@ describe('REST api v1.0, monitors resource:', function(){
   })
 
   it('try getting a report individually', function(done) {
-    superagent.get('http://localhost:3000/api/v1/monitors/' + id + '/reports/' + reports[0])
+    superagent.get(artifacts.url + '/api/v1/monitors/' + id + '/reports/' + reports[0])
       .end(function(e,res){
         expect(e).to.eql(null);
         expect(res.status).to.eql(200);
@@ -100,7 +106,7 @@ describe('REST api v1.0, monitors resource:', function(){
   })
 
   it('try to post a report and overwrite monitors_id (should fail)', function(done) {
-    superagent.post('http://localhost:3000/api/v1/monitors/' + id + '/reports')
+    superagent.post(artifacts.url + '/api/v1/monitors/' + id + '/reports')
       .send({
         value: 'BOGUS',
         monitors_id: 'HAXORS'
@@ -114,7 +120,7 @@ describe('REST api v1.0, monitors resource:', function(){
 
   it('delete the reports', function(done) {
   	_.forEach( reports, function( report_id ) {
-	  	superagent.del('http://localhost:3000/api/v1/monitors/' + id + '/reports/'+ report_id)
+	  	superagent.del(artifacts.url + '/api/v1/monitors/' + id + '/reports/'+ report_id)
 	  		.end(function(e,res){
 	  			expect(e).to.eql(null);
 	  			expect(res.status).to.eql(200);
@@ -125,7 +131,7 @@ describe('REST api v1.0, monitors resource:', function(){
   	});
   })
   it('get the empty reports list', function(done) {
-  	superagent.get('http://localhost:3000/api/v1/monitors/' + id + '/reports')
+  	superagent.get(artifacts.url + '/api/v1/monitors/' + id + '/reports')
   		.end(function(e,res){
   			expect(e).to.eql(null);
   			expect(res.status).to.eql(200);
@@ -137,7 +143,8 @@ describe('REST api v1.0, monitors resource:', function(){
   })
 
   it('delete the monitor', function(done){
-  	superagent.del('http://localhost:3000/api/v1/monitors/' + id)
+  	superagent.del(artifacts.url + '/api/v1/monitors/' + id)
+      .set('Authorization', 'Bearer ' + artifacts.token)
   		.end(function(e,res){
   			expect(e).to.eql(null);
   			expect(res.status).to.eql(200);
@@ -146,7 +153,8 @@ describe('REST api v1.0, monitors resource:', function(){
   })
 
   it('fetch the non-existent deleted monitor (should fail)', function(done){
-  	superagent.get('http://localhost:3000/api/v1/monitors/' + id)
+  	superagent.get(artifacts.url + '/api/v1/monitors/' + id)
+      .set('Authorization', 'Bearer ' + artifacts.token)
   		.end(function(e,res){
   			expect(e).to.eql(null);
   			expect(res.status).to.eql(404);

--- a/test/test-artifacts.js
+++ b/test/test-artifacts.js
@@ -1,0 +1,11 @@
+var Artifacts = function() {
+	this.port = 3000 + Math.floor(Math.random() * 1000);
+	this.dbName = 'test-' + this.port;
+	this.db = 'mongodb://localhost:27017/' + this.dbName;
+	this.username = 'test.user@example.com';
+	this.password = 'l33tp455w0rd!';
+	this.url = 'http://localhost:' + this.port;
+	this.token = null;
+};
+
+module.exports = new Artifacts();

